wikipedia_cleanup/predictor.py

```diff
@@ -82,46 +80,57 @@ def __init__(self, p: float = 0.5):
         self.p = p
 
     def predict_timeframe(
-        self,
-        data_key: np.ndarray,
-        additional_data: np.ndarray,
-        columns: List[str],
-        first_day_to_predict: date,
-        timeframe: int,
+            self,
+            data_key: np.ndarray,
+            additional_data: np.ndarray,
+            columns: List[str],
+            first_day_to_predict: date,
+            timeframe: int,
     ) -> bool:
         return random.random() <= self.p
 
 
-class MeanPredictor(ZeroPredictor):
+class RegressionPredictor(Predictor, ABC):
+    def __init__(self):
+        super().__init__()
+        self.last_known_key = None
+        self.last_known_timestamp = None
+        self.last_known_prediction = False
+
     def predict_timeframe(
-        self,
-        data_key: np.ndarray,
-        additional_data: np.ndarray,
-        columns: List[str],
-        first_day_to_predict: date,
-        timeframe: int,
+            self,
+            data_key: np.ndarray,
+            additional_data: np.ndarray,
+            columns: List[str],
+            first_day_to_predict: date,
+            timeframe: int,
     ) -> bool:
-        col_idx = columns.index("value_valid_from")
-        pred = self.next_change(data_key[:, col_idx])
-        if pred is None:
+        if not self._should_make_prediction(data_key, columns):
             return False
-        return (
-            first_day_to_predict <= pred <= first_day_to_predict + timedelta(timeframe)
+        key_column = columns.index("key")
+        value_valid_from_idx = columns.index("value_valid_from")
+        current_key = data_key[0, key_column]
+        if current_key == self.last_known_key:
```

Comment:
I am not so sure if this completely works. This is intended to be a cache. I think, we should cache the result of all predictors and not the boolean. Since the most inner loop is the `timeframe` this would result in false predictions. e.g. The prediction is in 5 days. The dayly prediction is False. Thus `self.last_known_prediction` is False. The next prediction will be weekly and we return False again. Even though the condition would now evaluate to True. I hope this is somehow understandable

Reply:
Changed it to store the date and recalculate the bool, should fix it